src/Cms/LanguageRoutes.php

```diff
@@ -29,9 +30,26 @@ public static function create(App $kirby): array
 				'pattern' => $language->pattern(),
 				'method'  => 'ALL',
 				'env'     => 'site',
-				'action'  => function ($path = null) use ($language) {
+				'action'  => function ($path = null) use ($kirby, $language) {
 					$result = $language->router()->call($path);
 
+					// redirect secondary-language pages that have
+					// been accessed with non-translated slugs in their path
+					// to their fully translated URL
+					if ($path !== null && $result instanceof Page) {
+						if (Str::endsWith($result->url(), $path) === false) {
+							$url   = $result->url();
+							$query = $kirby->request()->query()->toString();
+
+							// preserve query across redirect
+							if (empty($query) === false) {
+								$url .= '?' . $query;
+							}
+
+							return Response::redirect($url);
+						}
```

Comment:
`Question`: @bastianallgeier https://github.com/getkirby/kirby/pull/6601/files#diff-141beee66f7ef4aacf2cf10c723ab70676499ba12e2b2639fd82d9c4ce65e3ffL69-R114 tries to solve something similarly. Can you explain what? I don't think though we could move this new code to `::fallback` as `$result` actually is a Page object found. That's why I solved it here. (but maybe we could align the code style, e.g. `Response::redirect)` vs `$kirby->response()->redirect()`)

Reply:
That part tries to redirect `/de/something` to an absolute language URL by detecting the code. I'm not entirely sure anymore why we needed this. I think it's code that supported legacy behavior. 

I agree that we can't implement both it in the same place. I think the `$kirby->response()->redirect()` solution is a bit nicer when we already have access to the kirby instance.